### PR TITLE
LibPDF+LibGfx (in particular, JPEG): Use regular color codepath for jpeg data too

### PR DESCRIPTION
**Pull request by nico**
This handing the original jpeg channels to the color spaces, so this finally adds a way to get "raw" CMYK data from a jpeg, and also a way to know if jpeg data originally was grayscale.

The regula color codepath is fairly slow, so this makes pages with jpegs render subjectively quite a bit slower. Objectively, processing my 1000 file pdf test set goes from 45s to 1m57s. (But the majority of the time is spent on 0000849.pdf, a file with many many jpegs. If I filter that one file out in test_pdf.py, it's 45s -> 1m7s. Still 50% slower, but not 160% slower). In return, they're correct. And it provides additional incentive for making that codepath faster :^)

We now handle /Decode arrays on jpegs, and we now color-manage jpeg data in PDFs (including CMYK and grayscale jpegs).

Fixes all remaining color-related issues I'm aware of. (There are a bunch of missing non-color image-related issues still: We don't clip images, we ignore smask state on the graphics state, probably more.)

### CODE REVIEW

**Comment by nico** [comment]
[Color ](https://www.color.org/version4pdf.pdf), before:

Screenshot 2024-01-07 at 10 24 18 PM

After:

Screenshot 2024-01-07 at 10 23 37 PM

Several progressions on my nemesis 0000277.pdf. Before:

Screenshot 2024-01-07 at 10 24 39 PM

Screenshot 2024-01-07 at 10 24 47 PM

After (due to /Decode support):

Screenshot 2024-01-07 at 10 06 25 PM

Screenshot 2024-01-07 at 10 06 15 PM

0000277.pdf isn't the only file using /Decode to invert jpeg channels. Another before:

Screenshot 2024-01-07 at 10 27 58 PM

And after:

Screenshot 2024-01-07 at 10 27 28 PM

**Comment by LucasChollet** [comment]
First thing first, that's very exciting, and I'm really happy to see that we are now able to render that test correctly.

Buuuut, I'm really not a fan of having a separate way to get the CMYK bitmap from the decoder. That doesn't seem really scalable if we start to introduce new bitmaps like `12BitsRGB` or `16BitsARGB`.

That's not thoroughly though, but at least to propose something to talk about (let's bring that conversation on discord though, I put the first message here to dissuade a compulsive merge). What about having image decoder return a class like this:
```cpp
// This is supposed to be compatible~ish with ImageFrameDescriptor
class ExtendedBitmap {
	enum class Type {
		CMYK,
		WhateverNewVariantYouWant,
		ARGB32,
		Indexed // ? So we can split that out from the current Bitmap implementation.
	}
	bool has_juicy_stuff() { return m_juicy_bitmap; };
	RefPtr<BaseBitmap> juicy_image() { return m_juicy_bitmap; };
	RefPtr<ARGB32Bitmap> image() { // As you said, most clients don't care (yet?), so let's keep a compatibility layer.
		return m_cached_argb_converted_from_juicy_bitmap_on_first_request;
	}

	RefPtr<BaseBitmap> m_juicy_bitmap; // With all Bitmap variants enumerated in Type inheriting from BaseBitmap.
}

**Comment by nico** [comment]
> That doesn't seem really scalable if we start to introduce new bitmaps like 12BitsRGB or 16BitsARGB.

Bit depth is a different dimension than channel arrangement in my mind. For 16-bits argb, making Bitmap templatized on the channel type could work. Not saying that's necessarily the best design, but it seems alright at first blush. And 12-bit could just be stored as Bitmap<u16> then.

Your point still stands with CIELAB I suppose.

But you kind of need a new type to express the new channel type – cmyk bitmaps for the most part really are different from RGB bitmaps (which really are different from CIELAB bitmaps). The `scanline()` function has to return a different type.

The common bits are "2d byte buffer", which could contain the IPC things and volatileness and hidpi flag, but even things like image resizing kind of depend on channel config (are the channels linear?). So I'm not sure how useful a common super type is.

When drawing to a screen, you always want an (RGB) Bitmap, so I think it's good if the base API for that doesn't change. Most clients won't care about non-RGB.

Sophisticated clients like PixelPaint probably want to have their own internal Bitmap format (always 16-bit linear for example).

In summary, it seems complicated enough that I didn't want to deal with it too much and did the simple thing for now. Most of the changes in here are independent of that discussion anyways, and it's reasonably easy to change later. So I figured let's punt on figuring out the final approach. Maybe the right way is easier to see once we have some more multi color space code.

(Also, the current approach is in line with the existing vector image code.)

**Comment by nico** [comment]
I have a local branch that optimizes color space conversion a bit, and with those changes time goes for no profiles for jpegs at 45s to time for profiles for jpegs at 49s (instead of 1m7s above) without 0000849.pdf and at 56s (instead of 1m57s above) with it. Flipping through 0000849.pdf still doesn't feel _fast_, but it's a little more bearable. (And there's still lots of future room for optimization.) I can upload that branch as soon as this one is in.

**Review by LucasChollet — APPROVED** [review]
For reference, the discussion continued here:
https://discord.com/channels/830522505605283862/859531354219872266/1194090078142742588

I still do not like the modification to the API, but everything else looks good :^)
